src/Java.Interop.Tools.JavaTypeSystem/JavaModels/JavaMethodModel.cs

```diff
@@ -0,0 +1,188 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Java.Interop.Tools.JavaTypeSystem.Models
+{
+	public class JavaMethodModel : JavaMemberModel
+	{
+		public string Return { get; }
+		public string ReturnGeneric { get; }
+		public bool IsAbstract { get; }
+		public bool IsBridge { get; }
+		public string ReturnJni { get; }
+		public bool IsSynthetic { get; }
+		public bool IsSynchronized { get; }
+		public bool IsNative { get; }
+		public bool ReturnNotNull { get; }
+
+		public JavaTypeParameters TypeParameters { get; }
+		public JavaTypeReference? ReturnTypeModel { get; private set; }
+		public JavaMethodModel? BaseMethod { get; set; }
+		public List<JavaParameterModel> Parameters { get; } = new List<JavaParameterModel> ();
+		public List<JavaExceptionModel> Exceptions { get; } = new List<JavaExceptionModel> ();
+
+		public JavaMethodModel (string javaName, string javaVisibility, bool javaAbstract, bool javaFinal, bool javaStatic, string javaReturn, JavaTypeModel javaParentType, string deprecated, string jniSignature, bool isSynthetic, bool isBridge, string returnJni, bool isNative, bool isSynchronized, bool returnNotNull)
+			: base (javaName, javaStatic, javaFinal, javaVisibility, javaParentType, deprecated, jniSignature)
+		{
+			IsAbstract = javaAbstract;
+			Return = javaReturn;
+			ReturnGeneric = javaReturn;
+			IsBridge = isBridge;
+			IsSynthetic = isSynthetic;
+			ReturnJni = returnJni;
+			IsNative = isNative;
+			IsSynchronized = isSynchronized;
+			ReturnNotNull = returnNotNull;
+
+			TypeParameters = new JavaTypeParameters (this);
+		}
+
+		public override void Resolve (JavaTypeCollection types, List<JavaUnresolvableModel> unresolvables)
+		{
+			if (Name.Contains ('$')) {
+				unresolvables.Add (new JavaUnresolvableModel (this, "$", UnresolvableType.DollarSign));
+				return;
+			}
+
+			var type_parameters = GetApplicableTypeParameters ().ToArray ();
+
+			try {
+				ReturnTypeModel = types.ResolveTypeReference (Return, type_parameters);
+			} catch (JavaTypeResolutionException) {
+				unresolvables.Add (new JavaUnresolvableModel (this, Return, UnresolvableType.ReturnType));
+
+				return;
+			}
+
+			foreach (var p in Parameters.OfType<JavaParameterModel> ())
+				p.Resolve (types, unresolvables);
+		}
+
+		// Return method's type parameters, plus type parameters for any parent type(s).
+		public IEnumerable<JavaTypeParameter> GetApplicableTypeParameters ()
+		{
+			foreach (var jtp in TypeParameters)
+				yield return jtp;
+
+			if (ParentType != null)
+				foreach (var jtp in ParentType.GetApplicableTypeParameters ())
+					yield return jtp;
+		}
+
+		public void FindBaseMethod (JavaClassModel? type)
```

Comment:
Should this not be part of `Resolve()`?

Reply:
We do a 2-phase approach to ensure we have resolved all base types first, so we know where to look for base methods.

https://github.com/xamarin/java.interop/pull/849/files#diff-e9952181b160e4d6b3bebe013ced311fa85bfca566a889e26c3da85bdf59274eR169-R171